File: 700/752 - Open the Lock.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

class Solution {
    public int openLock(String[] deadends, String target) {
        int ret = 0;

        Set<String> visited = new HashSet<>();
        for (String s : deadends) {
            visited.add(s);
        }
        
        if (visited.contains("0000")) {
            return -1;
        } else if (target.equals("0000")) {
            return 0;
        }
        Queue<String> q = new LinkedList<>();
        
        visited.add("0000");
        q.offer("0000");
        
        while (!q.isEmpty()) {
            ++ret;

            int count = q.size();
            while (count-- > 0) {
                String curr = q.poll();
                
                for (int i = 0; i < 4; ++i) {
                    char c = curr.charAt(i);
                    String a = curr.substring(0, i), b = curr.substring(i + 1);

                    String prev = a + (char) (c == '0' ? '9' : c - 1) + b;
                    String next = a + (char) (c == '9' ? '0' : c + 1) + b;
                    if (prev.equals(target) || next.equals(target)) {
                        return ret;
                    }
                    
                    if (!visited.contains(prev)) {
                        visited.add(prev);
                        q.offer(prev);
                    }
                    
                    if (!visited.contains(next)) {
                        visited.add(next);
                        q.offer(next);
                    }
                }
            }
        }
        
        return -1;
    }
}

File: csharp/752.open-the-lock.cs
/*
 * @lc app=leetcode id=752 lang=csharp
 *
 * [752] Open the Lock
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {
    public int OpenLock(string[] deadends, string target) {
        var dead = new HashSet<string>(deadends);
        if (dead.Contains("0000")) {
            return -1;
        }
        if (target == "0000") {
            return 0;
        }
        int ret = 0;
        Queue<string> queue = new Queue<string>();
        queue.Enqueue("0000");
        dead.Add("0000");
        while (queue.Count > 0) {
            int count = queue.Count;
            ++ret;
            while (count-- > 0) {
                var curr = queue.Dequeue().ToCharArray();
                for (int i = 0; i < 4; ++i) {
                    for (int d = -1; d < 2; d += 2) {
                        char tmp = curr[i];
                        curr[i] = (char) (((int) curr[i] - (int) '0' + 10 + d) % 10 + (int) '0');
                        string next = new string(curr);
                        if (next == target) {
                            return ret;
                        }
                        if (!dead.Contains(next)) {
                            queue.Enqueue(next);
                            dead.Add(next);
                        }
                        curr[i] = tmp;
                    }
                }
            }
        }
        return -1;
    }
}
// @lc code=end

